### PR TITLE
feat(rt): OpenTelemetry based telemetry provider

### PR DESCRIPTION
**Pull request by aajtodd**
## Issue \#

N/A

## Description of changes

Add an (experimental) OpenTelemetry based `TelemetryProvider`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/AttributeUtils.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+internal fun <T> AttributeKey<T>.otelAttrKeyOrNull(value: T): OtelAttributeKey<T>? {
+    val otelKey = when (value) {
+        is String -> OtelAttributeKey.stringKey(name)
+        is Long -> OtelAttributeKey.longKey(name)
+        is Boolean -> OtelAttributeKey.booleanKey(name)
+        is Double -> OtelAttributeKey.doubleKey(name)
+        is List<*> -> {
+            when (value.firstOrNull()) {
```
Comment (lauzadis) [inline runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/AttributeUtils.kt:37]: This won't work if it's a mixed type `List<Any>`, is that ok?

**Review by lauzadis — APPROVED** [review]

File: runtime/observability/telemetry-api/common/src/aws/smithy/kotlin/runtime/telemetry/trace/CoroutineContextTraceExt.kt [hunk 2]
```diff
@@ -63,7 +65,11 @@ public suspend inline fun <R> withSpan(
     context: CoroutineContext = EmptyCoroutineContext,
     crossinline block: suspend CoroutineScope.(span: TraceSpan) -> R,
 ): R = try {
-    withContext(context + TraceSpanContext(span)) {
+    // after a span is created the current telemetry context may be updated, ensure the new coroutine gets the latest
+    // or else traces may be disconnected from their parent
+    val updatedCtx = coroutineContext[TelemetryProviderContext]?.provider?.contextManager?.current()
+    val telemetryCtxElement = (updatedCtx?.let { TelemetryContextElement(it) } ?: coroutineContext[TelemetryContextElement]) ?: EmptyCoroutineContext
```
Comment (ianbotsf) [inline runtime/observability/telemetry-api/common/src/aws/smithy/kotlin/runtime/telemetry/trace/CoroutineContextTraceExt.kt:71]: **Nit**: Parentheses unnecessary

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/AttributeUtils.kt [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+package aws.smithy.kotlin.runtime.telemetry.otel
+
+import aws.smithy.kotlin.runtime.util.AttributeKey
+import aws.smithy.kotlin.runtime.util.Attributes
+import aws.smithy.kotlin.runtime.util.get
+import io.opentelemetry.api.common.AttributeKey as OtelAttributeKey
+import io.opentelemetry.api.common.Attributes as OtelAttributes
+
+@Suppress("UNCHECKED_CAST")
+internal fun Attributes.toOtelAttributes(): OtelAttributes {
```
Comment (ianbotsf) [inline runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/AttributeUtils.kt:15]: **Nit**: Narrow suppression to minimal scope:

```kotlin
@Suppress("UNCHECKED_CAST")
val key = it as AttributeKey<Any>
```

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+import io.opentelemetry.api.metrics.LongUpDownCounter as OtelLongUpDownCounter
+import io.opentelemetry.api.metrics.Meter as OpenTelemetryMeter
+
+internal class OtelMeterProvider(private val otel: OpenTelemetry) : MeterProvider {
+    override fun getOrCreateMeter(scope: String, attributes: Attributes): Meter {
+        val meter = otel.getMeter(scope)
+        // FIXME: meter level attributes not supported yet by Java impl
+        // https://github.com/open-telemetry/opentelemetry-java/issues/5503
+        return OtelMeter(meter)
+    }
```
Comment (ianbotsf) [inline runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:26]: **Question**: Should we warn or throw an exception if non-empty attributes are passed?

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    private val otelMeter: OpenTelemetryMeter,
+) : Meter {
+    override fun createUpDownCounter(name: String, units: String?, description: String?): UpDownCounter {
+        val counter = otelMeter.upDownCounterBuilder(name)
+            .apply {
+                description?.let { setDescription(it) }
+                units?.let { setUnit(units) }
+            }
+            .build()
+        return OtelUpDownCounterImpl(counter)
```
Comment (ianbotsf) [inline runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:39]: **Question**: Are the `?.let` null-checks necessary? The property values on new builders should already be `null` so re-setting them to `null` shouldn't have any negative effect:

```kotlin
val counter = otelMeter
    .upDownCounterBuilder(name)
    .setDescription(description)
    .setUnit(units)
    .build()
return OtelUpDownCounterImpl(counter)
```

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelTracerProvider.kt [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+import io.opentelemetry.api.trace.SpanKind as OtelSpanKind
+import io.opentelemetry.api.trace.StatusCode as OtelStatus
+import io.opentelemetry.api.trace.Tracer as OtelTracer
+
+internal class OtelTracerProvider(private val otel: OpenTelemetry) : TracerProvider {
+    override fun getOrCreateTracer(scope: String, attributes: Attributes): Tracer =
+        OtelTracer(otel.getTracer(scope), attributes)
+}
+
+private class OtelTracer(
```
Comment (ianbotsf) [inline runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelTracerProvider.kt:23]: **Suggestion**: This class has the same name as an import alias. I suggest changing it to `OtelTracerAdapter` or `OtelTracerImpl` or something.

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelTracerProvider.kt [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+                if (!tracerAttributes.isEmpty) {
+                    setAllAttributes(initialAttributes.toOtelAttributes())
+                }
+
+                if (!initialAttributes.isEmpty) {
+                    setAllAttributes(initialAttributes.toOtelAttributes())
+                }
+            }
+            .setSpanKind(spanKind.toOtelSpanKind())
```
Comment (ianbotsf) [inline runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelTracerProvider.kt:49]: **Question**: Same question as above about whether null/empty checks are necessary.

File: runtime/observability/telemetry-provider-otel/jvm/test/aws/smithy/kotlin/runtime/telemetry/otel/AttributeTests.kt [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+        val longAttr = AttributeKey<Long>("key")
+        val otelLongKey = longAttr.otelAttrKeyOrNull(0L)
+        assertNotNull(otelLongKey)
+        assertIs<OtelAttributeKey<Long>>(otelLongKey)
+
+        val boolAttr = AttributeKey<Boolean>("key")
+        val otelBoolKey = boolAttr.otelAttrKeyOrNull(true)
+        assertNotNull(otelBoolKey)
+        assertIs<OtelAttributeKey<Long>>(otelBoolKey)
```
Comment (ianbotsf) [inline runtime/observability/telemetry-provider-otel/jvm/test/aws/smithy/kotlin/runtime/telemetry/otel/AttributeTests.kt:32]: **Nit**: The generic type parameter checks are meaningless in these `assertIs` calls because of type erasure (as evidenced by the fact that the presumably-erroneous `assertIs<OtelAttributeKey<Long>>(otelBoolKey)` check passes). Without reflection, I think the best we can assert is `OtelAttributeKey<*>`.

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/OperationTelemetry.kt [hunk 9]
```diff
@@ -79,5 +72,12 @@ internal fun<I, O> SdkHttpOperation<I, O>.instrument(): Pair<TraceSpan, Coroutin
         telemetry.spanKind,
         parentCtx,
     )
+
+    val telemetryCtx = TelemetryProviderContext(telemetry.provider) +
+        LoggingContextElement(
+            "rpc" to rpcName,
+            "sdkInvocationId" to context.sdkInvocationId,
+        )
```
Comment (ianbotsf) [inline runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/OperationTelemetry.kt:80]: **Question**: In the previous code (deleted above), this new `telemetryCtx` also included `TelemetryContextElement(parentCtx)`. Why is that removed now?

**Review by ianbotsf — APPROVED** [review]
Nice, seems like the wrapping layer is pretty minimal.

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/AttributeUtils.kt (hunk 1 above)
Comment (aajtodd) [reply runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/AttributeUtils.kt:37]: Yeah it's fine we are in control of what attributes we are setting, if we are setting `List<Any>` we are doing something wrong probably

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt (hunk 4 above)
Comment (aajtodd) [reply runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:26]: Part of me is thinking we just remove `attributes` from the `getOrCreate` level interfaces (`MeterProvider` and `TracerProvider`).

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelTracerProvider.kt (hunk 7 above)
Comment (aajtodd) [reply runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelTracerProvider.kt:49]: avoids unnecessary empty map allocations

File: runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt (hunk 5 above)
Comment (aajtodd) [reply runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:39]: They are:

```
w: file:///smithy-kotlin/runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:38:29 Type mismatch: inferred type is String? but String was expected
```

Comment (aajtodd) [reply runtime/observability/telemetry-provider-otel/jvm/src/aws/smithy/kotlin/runtime/telemetry/otel/OtelMeterProvider.kt:39]: well at least if you want to get past warnings as errors check

File: runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/OperationTelemetry.kt (hunk 9 above)
Comment (aajtodd) [reply runtime/protocol/http-client/common/src/aws/smithy/kotlin/runtime/http/operation/OperationTelemetry.kt:80]: It's now set in the coroutine [extensions](https://github.com/awslabs/smithy-kotlin/blob/562508c799ffff7e257c6d6818fd3a91c22efcb3/runtime/observability/telemetry-api/common/src/aws/smithy/kotlin/runtime/telemetry/trace/CoroutineContextTraceExt.kt#L70-L71) so that it gets set at the correct place/time by default.

File: runtime/observability/telemetry-provider-otel/jvm/test/aws/smithy/kotlin/runtime/telemetry/otel/AttributeTests.kt (hunk 8 above)
Comment (aajtodd) [reply runtime/observability/telemetry-provider-otel/jvm/test/aws/smithy/kotlin/runtime/telemetry/otel/AttributeTests.kt:32]: 🤦
